File: src/utils/learningRate.js
/**
 * 학습률 계산 유틸리티
 *
 * 학습률: 0 ~ 100 (퍼센트)
 * 상태 그룹:
 *   - 어려워요 (Difficult): 0 ~ 39  → 빨간색
 *   - 학습 중 (Learning):  40 ~ 79  → 파란색
 *   - 외웠어요 (Memorized): 80 ~ 100 → 초록색
 */

// --- 상태 그룹 상수 ---
export const LEARNING_STATUS = {
  DIFFICULT: 'difficult',   // 어려워요
  LEARNING: 'learning',     // 학습 중
  MEMORIZED: 'memorized',   // 외웠어요
};

export const LEARNING_STATUS_CONFIG = {
  [LEARNING_STATUS.DIFFICULT]: {
    label: '어려워요',
    color: '#EF4444',       // red-500
    bgColor: 'bg-red-50',
    textColor: 'text-red-600',
    borderColor: 'border-red-200',
    dotColor: 'bg-red-500',
    range: [0, 39],
  },
  [LEARNING_STATUS.LEARNING]: {
    label: '학습 중',
    color: '#3B82F6',       // blue-500
    bgColor: 'bg-blue-50',
    textColor: 'text-blue-600',
    borderColor: 'border-blue-200',
    dotColor: 'bg-blue-500',
    range: [40, 79],
  },
  [LEARNING_STATUS.MEMORIZED]: {
    label: '외웠어요',
    color: '#22C55E',       // green-500
    bgColor: 'bg-green-50',
    textColor: 'text-green-600',
    borderColor: 'border-green-200',
    dotColor: 'bg-green-500',
    range: [80, 100],
  },
};

// --- 퀴즈 유형별 기본 보상/감소량 ---
const QUIZ_TYPE_WEIGHT = {
  multiple: 1.0,      // 객관식: 기본
  short: 1.4,         // 주관식: 40% 더 높은 보상
  'toefl-complete': 1.8, // Complete-the-Word: 80% 더 높은 보상
};

// --- 학습률 변동 상수 ---
const BASE_CORRECT_GAIN = 12;           // 정답 시 기본 상승량
const FIRST_WRONG_PENALTY = -5;          // 첫 오답: 소량 감소 (입력 실수 감안)
const REPEATED_WRONG_PENALTY = -10;      // 반복 오답: 더 큰 감소
const MAX_WRONG_MULTIPLIER = 3;          // 최대 오답 배수
const REASKED_CORRECT_RECOVERY = 0.6;    // 재출제 정답: 감소분의 60% 복구
const AI_SIMILAR_CORRECT_RECOVERY = 1.0; // AI 유사 문제 정답: 전부 복구 + 소량 향상
const AI_SIMILAR_BONUS = 3;              // AI 유사 문제 추가 보너스

/**
 * 학습률에서 상태 그룹 반환
 * @param {number} rate - 학습률 (0-100)
 * @returns {string} - LEARNING_STATUS 값
 */
export function getLearningStatus(rate) {
  const r = typeof rate === 'number' ? rate : 0;
  if (r >= 80) return LEARNING_STATUS.MEMORIZED;
  if (r >= 40) return LEARNING_STATUS.LEARNING;
  return LEARNING_STATUS.DIFFICULT;
}

/**
 * 학습률에 해당하는 색상 반환 (도넛 그래프용)
 * 부드러운 그라데이션을 위해 구간 내 보간
 * @param {number} rate - 학습률 (0-100)
 * @returns {string} - CSS 색상 값
 */
export function getLearningRateColor(rate) {
  const r = Math.max(0, Math.min(100, rate || 0));

  if (r < 40) {
    // 빨간색 → 파란색 전환 (0~39)
    const t = r / 39;
    return interpolateColor('#EF4444', '#3B82F6', t);
  }
  if (r < 80) {
    // 파란색 → 초록색 전환 (40~79)
    const t = (r - 40) / 39;
    return interpolateColor('#3B82F6', '#22C55E', t);
  }
  // 초록색 (80~100)
  return '#22C55E';
}

/**
 * 두 hex 색상 사이 보간
 */
function interpolateColor(color1, color2, t) {
  const r1 = parseInt(color1.slice(1, 3), 16);
  const g1 = parseInt(color1.slice(3, 5), 16);
  const b1 = parseInt(color1.slice(5, 7), 16);
  const r2 = parseInt(color2.slice(1, 3), 16);
  const g2 = parseInt(color2.slice(3, 5), 16);
  const b2 = parseInt(color2.slice(5, 7), 16);

  const r = Math.round(r1 + (r2 - r1) * t);
  const g = Math.round(g1 + (g2 - g1) * t);
  const b = Math.round(b1 + (b2 - b1) * t);

  return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
}

/**
 * 정답 시 학습률 변동 계산
 *
 * @param {object} params
 * @param {number} params.currentRate - 현재 학습률
 * @param {string} params.quizType - 'multiple' | 'short' | 'toefl-complete'
 * @param {boolean} params.isReasked - 이 문제가 재출제된 것인지 (같은 세션에서 틀린 후)
 * @param {boolean} params.isAiSimilar - AI 기반 유사 문제인지
 * @param {number} params.lastPenalty - 직전 오답 시 감소했던 양 (복구 계산용)
 * @returns {number} - 새로운 학습률
 */
export function calculateCorrectRate({
  currentRate = 0,
  quizType = 'multiple',
  isReasked = false,
  isAiSimilar = false,
  lastPenalty = 0,
}) {
  const weight = QUIZ_TYPE_WEIGHT[quizType] || 1.0;
  let gain;

  if (isAiSimilar && lastPenalty > 0) {
    // AI 유사 문제 정답: 감소분 전부 복구 + 소량 보너스
    gain = Math.abs(lastPenalty) * AI_SIMILAR_CORRECT_RECOVERY + AI_SIMILAR_BONUS;
  } else if (isReasked && lastPenalty > 0) {
    // 재출제 정답: 감소분의 일부 복구
    gain = Math.abs(lastPenalty) * REASKED_CORRECT_RECOVERY;
  } else {
    // 일반 정답
    gain = BASE_CORRECT_GAIN * weight;
  }

  return clampRate(currentRate + gain);
}

/**
 * 오답 시 학습률 변동 계산
 *
 * @param {object} params
 * @param {number} params.currentRate - 현재 학습률
 * @param {number} params.wrongCount - 이 단어의 누적 오답 횟수 (이번 오답 포함 전 값)
 * @returns {{ newRate: number, penalty: number }} - 새로운 학습률과 감소량
 */
export function calculateWrongRate({ currentRate = 0, wrongCount = 0 }) {
  let penalty;

  if (wrongCount === 0) {
    // 첫 오답: 소량 감소 (입력 실수, 깜빡하기 감안)
    penalty = FIRST_WRONG_PENALTY;
  } else {
    // 반복 오답: 횟수에 비례해 더 큰 감소
    const multiplier = Math.min(wrongCount, MAX_WRONG_MULTIPLIER);
    penalty = REPEATED_WRONG_PENALTY * multiplier;
  }

  const newRate = clampRate(currentRate + penalty);
  return { newRate, penalty };
}

/**
 * 학습률을 0-100 범위로 제한
 */
function clampRate(rate) {
  return Math.max(0, Math.min(100, Math.round(rate)));
}

/**
 * 단어 목록을 학습 상태별로 그룹핑
 * @param {Array} words - 단어 배열 (각각 learningRate 필드 보유)
 * @returns {Object} - { difficult: [], learning: [], memorized: [] }
 */
export function groupWordsByStatus(words) {
  const groups = {
    [LEARNING_STATUS.DIFFICULT]: [],
    [LEARNING_STATUS.LEARNING]: [],
    [LEARNING_STATUS.MEMORIZED]: [],
  };

  words.forEach(word => {
    const status = getLearningStatus(word.learningRate);
    groups[status].push(word);
  });

  return groups;
}

/**
 * 단어 목록을 학습률 기준으로 정렬
 * @param {Array} words - 단어 배열
 * @param {'asc' | 'desc'} order - 정렬 방향
 * @returns {Array} - 정렬된 단어 배열
 */
export function sortByLearningRate(words, order = 'asc') {
  return [...words].sort((a, b) => {
    const rateA = a.learningRate || 0;
    const rateB = b.learningRate || 0;
    return order === 'asc' ? rateA - rateB : rateB - rateA;
  });
}

File: ios/VocaLoop/Core/Models/LearningRate.swift
import SwiftUI

/// 웹 `src/utils/learningRate.js`의 이식.
///
/// 중요: 학습 상태는 서버의 `status` 필드가 아니라 **학습률에서 파생**한다.
/// 웹이 그렇게 하고 있고, 두 클라이언트가 같은 단어를 다르게 분류하면 안 된다.
enum LearningStatus: String, CaseIterable, Sendable {
    /// 0 ~ 39
    case difficult
    /// 40 ~ 79
    case learning
    /// 80 ~ 100
    case memorized

    init(rate: Int) {
        switch rate {
        case 80...: self = .memorized
        case 40...: self = .learning
        default: self = .difficult
        }
    }

    var label: String {
        switch self {
        case .difficult: return "어려워요"
        case .learning: return "학습 중"
        case .memorized: return "외웠어요"
        }
    }

    /// 목록 그룹 머리의 타일 기호.
    ///
    /// 셋이 한 흐름으로 읽혀야 한다 — 아직 모름 → 오르는 중 → 끝.
    /// 탭바가 쓰는 기호(book.closed·brain·arrow.trianglehead.clockwise·gearshape)는
    /// 뜻이 겹치므로 피한다.
    var symbolName: String {
        switch self {
        case .difficult: return "questionmark"
        case .learning: return "arrow.up.right"
        case .memorized: return "checkmark"
        }
    }

    // 웹은 같은 상태를 두 곳에서 서로 다른 팔레트로 그린다. 헷갈리기 쉬우니 나눠 둔다.
    //  - 목록 그룹 헤더(`LEARNING_STATUS_CONFIG`): Tailwind raw red/blue/green
    //  - 상태 배지(`LearningStatusBadge`): 디자인 토큰 danger/brand/success
    // 초록만 값이 다르다 (green-600 #16A34A vs success-600 #059669).

    /// 목록 그룹 헤더의 점 — red/blue/green-500.
    var dotColor: Color {
        switch self {
        case .difficult: return Color(hex: 0xEF4444)
        case .learning: return Color(hex: 0x3B82F6)
        case .memorized: return Color(hex: 0x22C55E)
        }
    }

    /// 목록 그룹 헤더의 글자 — red/blue/green-600.
    var textColor: Color {
        switch self {
        case .difficult: return Color.adaptive(light: 0xDC2626, dark: 0xFCA5A5)
        case .learning: return Color.adaptive(light: 0x2563EB, dark: 0x93C5FD)
        case .memorized: return Color.adaptive(light: 0x16A34A, dark: 0x86EFAC)
        }
    }

    /// 상태 배지의 점 — danger/brand/success-500.
    var badgeDotColor: Color {
        switch self {
        case .difficult: return Color(hex: 0xEF4444)
        case .learning: return Color(hex: 0x3B82F6)
        case .memorized: return Color(hex: 0x10B981)
        }
    }

    /// 상태 배지의 글자 — danger/brand/success-600.
    var badgeTextColor: Color {
        switch self {
        case .difficult: return Color.adaptive(light: 0xDC2626, dark: 0xFCA5A5)
        case .learning: return Color.adaptive(light: 0x2563EB, dark: 0x93C5FD)
        case .memorized: return Color.adaptive(light: 0x059669, dark: 0xA7F3D0)
        }
    }

    /// 상태 배지의 배경 — danger/brand/success-50.
    var badgeBackgroundColor: Color {
        switch self {
        case .difficult: return Color.adaptive(light: 0xFEF2F2, dark: 0xEF4444, darkAlpha: 0.12)
        case .learning: return Color.adaptive(light: 0xEFF6FF, dark: 0x3B82F6, darkAlpha: 0.12)
        case .memorized: return Color.adaptive(light: 0xECFDF5, dark: 0x10B981, darkAlpha: 0.12)
        }
    }
}

enum LearningRate {
    /// 도넛 링 색. 구간 안에서 보간해 부드럽게 넘어간다.
    /// 0→39 빨강→파랑, 40→79 파랑→초록, 80+ 초록.
    static func color(for rate: Int) -> Color {
        let r = Double(max(0, min(100, rate)))

        if r < 40 {
            return interpolate(0xEF4444, 0x3B82F6, r / 39)
        }
        if r < 80 {
            return interpolate(0x3B82F6, 0x22C55E, (r - 40) / 39)
        }
        return Color(hex: 0x22C55E)
    }

    // MARK: - 학습률 변동
    //
    // 웹 `learningRate.js`의 상수를 그대로 쓴다. 값을 바꾸면 같은 계정의
    // 학습률이 웹에서 푸느냐 앱에서 푸느냐에 따라 달라진다.

    private static let baseCorrectGain = 12.0
    private static let firstWrongPenalty = -5.0
    private static let repeatedWrongPenalty = -10.0
    private static let maxWrongMultiplier = 3

    /// 퀴즈 유형별 보상 가중치. 웹 `QUIZ_TYPE_WEIGHT`와 값이 같아야 한다.
    static func weight(for stage: AdaptiveStage) -> Double {
        switch stage {
        case .flashcard: return 0.5      // 가벼운 확인
        case .multiple: return 1.0
        case .shortEnKo, .shortKoEn: return 1.4 // 40% 더 높은 보상
        case .completeWord: return 1.8   // 80% 더 높은 보상
        }
    }

    static func weight(for mode: QuizMode) -> Double { weight(for: mode.stage) }

    /// 정답 시 새 학습률.
    static func rateAfterCorrect(currentRate: Int, stage: AdaptiveStage) -> Int {
        clamp(Double(currentRate) + baseCorrectGain * weight(for: stage))
    }

    static func rateAfterCorrect(currentRate: Int, mode: QuizMode) -> Int {
        rateAfterCorrect(currentRate: currentRate, stage: mode.stage)
    }

    /// 오답 시 새 학습률. 반복해서 틀릴수록 더 크게 깎인다.
    static func rateAfterWrong(currentRate: Int, wrongCount: Int) -> Int {
        let penalty: Double = wrongCount == 0
            ? firstWrongPenalty
            : repeatedWrongPenalty * Double(min(wrongCount, maxWrongMultiplier))

        return clamp(Double(currentRate) + penalty)
    }

    private static func clamp(_ rate: Double) -> Int {
        max(0, min(100, Int(rate.rounded())))
    }

    /// 웹 `sortByLearningRate(words, 'asc')` — 학습률이 낮은 단어부터.
    ///
    /// 자바스크립트 `sort`는 안정 정렬이라 학습률이 같으면 원래 순서를 지킨다.
    /// Swift의 `sorted`는 그렇지 않아 색인을 함께 비교해 순서를 고정한다.
    static func sortedByRate(_ words: [Word]) -> [Word] {
        words.enumerated()
            .sorted {
                $0.element.learningRate == $1.element.learningRate
                    ? $0.offset < $1.offset
                    : $0.element.learningRate < $1.element.learningRate
            }
            .map(\.element)
    }

    static func interpolate(_ from: UInt32, _ to: UInt32, _ t: Double) -> Color {
        let t = max(0, min(1, t))
        let channel = { (shift: UInt32) -> CGFloat in
            let a = CGFloat((from >> shift) & 0xFF)
            let b = CGFloat((to >> shift) & 0xFF)
            return (a + (b - a) * t) / 255
        }

        // `Color(hex:)`와 같은 경로(UIColor)로 만들어야 두 색이 같은 값으로 비교된다.
        return Color(uiColor: UIColor(
            red: channel(16),
            green: channel(8),
            blue: channel(0),
            alpha: 1
        ))
    }
}

extension Word {
    /// 학습률에서 파생한 상태. 목록 그룹핑과 배지가 모두 이걸 쓴다.
    var learningStatus: LearningStatus {
        LearningStatus(rate: learningRate)
    }
}
